clang/lib/CIR/CodeGen/CIRGenFunction.h

```diff
@@ -1991,6 +1994,11 @@ class CIRGenFunction : public CIRGenTypeCache {
     mlir::Block *getEntryBlock() { return EntryBlock; }
 
     mlir::Location BeginLoc, EndLoc;
+    // Each SmallVector<APSInt> object is corresponding to a case region, empty
+    // vector means default case region.
```

Comment:
Do these 3 new members really need to be inside `LexicalScope`? Perhaps I missed something. Feels like they provide more details of switch lowering we'd want inside `LexicalScope`. Worst case they can be inside a struct declared inside `CIRGenStmt.cpp` and passed around over there? My impression is:

More specifically, `lastCaseBlock` could just be a local variable inside `buildCaseCompoundStmt`?

Reply:
Yes, it's a unnecessary change, let me have a explanation.

My initial (obsoleting) idea is `CaseStmt` may be inside any statements, so it's better to handle it in `buildStmt()` instead of `buildSwitchStmt()`. The logic would be like:
```
auto buildStmt(const Stmt *S) {
  ...
  if (S is CaseStmt)
    buildCaseStmt(S);
  ...
}
auto buildSwitchStmt(const SwitchStmt &S) {
  ...
  buildStmt(S.body());
  ...
}
auto buildCaseStmt(const CaseStmt &S) {
  ...
  auto switchLexScope = currLexScope;
  while (!switchLexScope->isSwitch())
    switchLexScope = switchLexScope->parrentLexScope;
  // Then update case attributes and `lastCaseBlock` in switchLexScope
  ...
}
```

So I moved some variables to `LexicalScope` to make `buildCaseStmt()` only has on parameter `const CaseStmt &S`.

Later, I found all I need to do is only handling the `CaseStmt` in `CompoundStmt`'s body, now that we wouldn't handle the scope cross in this pr. Then the previous change that defined variables inside `LexicalScope` looks weird. 
After thinking, I kept the change. Because we still need it in the future, when we want to support `CaseStmt` nested in other scope like `WhileStmt`.

Hope I explained it clearly, let me know if you still have concern on this change, I could revert it :）

Reply:
Thanks for the explanation, it make sense for the most part, though I still have few concerns.

Before we set on the final approach, one note is that we usually try to avoid visiting nodes multiple times. For example, `collectCompoundStmtHasCase` is being called before calling another method that is applying the same traversal. Since we can already walk scopes up to update information, why this is needed?

In most cases we do one visit and keep some information around (e.g. Lexscope) and that's usually enough.

Reply:
I'm afraid we have to visit the nodes twice, the problem is also related to scope.

Ideally, we should create a new scope for every `CompoundStmt`, all the problems will be gone.
However we haven't supported `CaseStmt` inside the new scope. So I add a function `buildCaseCompoundStmt()`, which won't create new scope.
If there is a `CaseStmt` inside the `CompoundStmt` subtree we call `buildCaseCompoundStmt()`, if not, call `buildStmt()` to create a new scope.

```
switch(a) {
// Should we create a new scope for case body stmt? It depends.
case 1: {
    {
      // There may be case here
      ...
    }
    Foo foo;
  }
}

switch(a) {
// Call buildStmt() and create a new scope,.
case 1: {
    {
      int x = 0;
    }
    Foo foo;
  }
}

switch(a) {
// Call buildCaseCompoundStmt(), don't create a new scope.
case 1: {
    {
      case 2:
    }
    Foo foo;
  }
}
```

So the logic will be like
```
buildCaseCompoundStmt(CompoundStmt &S) {
  for (auto *body : S.body()) {
    if (body is compound) {
      if (hasCase)
        buildCaseCompoundStmt(body);
      else
        // We can't call buildCaseCompoundStmt() here, otherwise we will lose the new scope.
        buildStmt(body, /*useCurrentScope=*/false);
    } else {
      // If we want to calculate hasCase inside buildCaseCompoundStmt,
      // we need to build this node before the parent node could decide whether should create a new scope.
      buildStmt(body, /*useCurrentScope=*/true);
    }
  }
}
```

So we can't compute `hasCase` inside `buildCaseCompoundStmt`, and extra collect function is needed.

Reply:
> I'm afraid we have to visit the nodes twice, the problem is also related to scope.

As general feedback, that much fine grained level of tracking + multiple visitors complicates the design and is usually an indicative that we're trying to do more work than necessary and/or that CIR design needs more improvements. You might have picked a complicated feature to work in this PR, but bear with me cause eventually we are going to converge on something. Also apologies for pushing harder into making this more comprehensive than initially needed.

> Ideally, we should create a new scope for every CompoundStmt, all the problems will be gone. However we haven't supported CaseStmt inside the new scope.

Somehow I missed that in previous reviews. Given the mentioned complexity, it's possible that we might have to change `cir.switch/cir.case` a bit to better support scopes as written. We can then later add extra passes that will clean up unncessary stuff.

As you pointed out (thanks for the nice examples), they are very complex to handle. In the end, the approach of the current PR is trying to detect simple cases involving nested compound statements, but they are just papering over all the inter scopes jumps that are necessary (e.g. https://godbolt.org/z/E65PE4vG1). I don't think we are ready to mess up with compound statements just yet, we need to build up on Oleg's work here. To make CIRGen flow natural, we'll need to create new scopes and to add synthetic forms of `cir.case` we can patch up later while flattening the CFG. Not much different from case statements in loops like we discussed before.

My suggestion is to take a step back here:
- Not touch compound statements in this PR (only if you are adding new NYI asserts, but no additional visitors).
- Tackle only cases that don't have extra compound statements, like in #520 and #521.

Once Oleg lands his work on gotos, I'm happy to help you continue the compound work and improve CIR to make it happen. How about that?

Reply:
It sounds reasonable, the compound logic has been reverted.
I keep the change of rewriting some function, though they wouldn't help to resolve #520 #521, I expect it's a better implementation as we discussed above :)

Reply:
Hi. I took another look at the patch and there are too many refactoring going on at the same time there's functionality change, and I'm not very confident to let this land right now. I'd like to first approve a PR that only does the NFC refactoring, and later on a PR that changes the minimum code that fixes only the problem in question. You could create a new PR for the refactoring and later update this one when it lands, your call on the PR approach.

Reply:
Sure, refactoring pr created https://github.com/llvm/clangir/pull/552, thanks for comments~